File: app/src/main/java/com/laushkina/anastasia/psychosophy/domain/Psychotype.java
package com.laushkina.anastasia.psychosophy.domain;

import android.os.Parcel;
import android.os.Parcelable;

import com.laushkina.anastasia.psychosophy.domain.test.Function;

public enum Psychotype implements Parcelable {
    Augustine(new Function[]{Function.Logic, Function.Emotion, Function.Physics, Function.Will}),
    Pascal(new Function[]{Function.Logic, Function.Emotion, Function.Will, Function.Physics}),
    Berthier(new Function[]{Function.Logic, Function.Physics, Function.Emotion, Function.Will}),
    Plato(new Function[]{Function.Logic, Function.Physics, Function.Will, Function.Emotion}),
    Lao(new Function[]{Function.Logic, Function.Will, Function.Physics, Function.Emotion}),
    Einstein(new Function[]{Function.Logic, Function.Will, Function.Emotion, Function.Physics}),

    Anderson(new Function[]{Function.Emotion, Function.Logic, Function.Will, Function.Physics}),
    Rousseau(new Function[]{Function.Emotion, Function.Logic, Function.Physics, Function.Will}),
    Bukharin(new Function[]{Function.Emotion, Function.Physics, Function.Logic, Function.Will}),
    Pushkin(new Function[]{Function.Emotion, Function.Physics, Function.Will, Function.Logic}),
    Ghazali(new Function[]{Function.Emotion, Function.Will, Function.Logic, Function.Physics}),
    Parsnip(new Function[]{Function.Emotion, Function.Will, Function.Physics, Function.Logic}),

    Aristippus(new Function[]{Function.Physics, Function.Logic, Function.Will, Function.Emotion}),
    Epicurus(new Function[]{Function.Physics, Function.Logic, Function.Emotion, Function.Will}),
    Borgia(new Function[]{Function.Physics, Function.Emotion, Function.Logic, Function.Will}),
    Dumas(new Function[]{Function.Physics, Function.Emotion, Function.Will, Function.Logic}),
    Goethe(new Function[]{Function.Physics, Function.Will, Function.Logic, Function.Emotion}),
    Chekhov(new Function[]{Function.Physics, Function.Will, Function.Emotion, Function.Logic}),

    Akhmatova(new Function[]{Function.Will, Function.Emotion, Function.Logic, Function.Physics}),
    Tolstoy(new Function[]{Function.Will, Function.Emotion, Function.Physics, Function.Logic}),
    Lenin(new Function[]{Function.Will, Function.Logic, Function.Physics, Function.Emotion}),
    Socrates(new Function[]{Function.Will, Function.Logic, Function.Emotion, Function.Physics}),
    Napoleon(new Function[]{Function.Will, Function.Physics, Function.Logic, Function.Emotion}),
    Twardowski(new Function[]{Function.Will, Function.Physics, Function.Emotion, Function.Logic});

    private Function[] functions;

    Psychotype(Function[] functions){
        this.functions = functions;
    }

    static Psychotype resolve(Parcel in) {
        return valueOf(in.readString());
    }

    public boolean areFunctionsEqual(Function[] functions){
        return functions.length == 4 &&
                this.functions[0] == functions[0] &&
                this.functions[1] == functions[1] &&
                this.functions[2] == functions[2] &&
                this.functions[3] == functions[3];
    }

    public static Psychotype resolve(Function[] functions){
        Psychotype foundType = null;
        for (Psychotype psychotype: values()) {
            if (psychotype.areFunctionsEqual(functions)) {
                foundType = psychotype;
            }
        }
        return foundType;
    }

    public Function[] getFunctions(){
        return functions;
    }


    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(name());
    }

    @Override
    public int describeContents() {
        return 0;
    }

    public static final Creator<Psychotype> CREATOR = new Creator<Psychotype>() {
        @Override
        public Psychotype createFromParcel(Parcel in) {
            return Psychotype.resolve(in);
        }

        @Override
        public Psychotype[] newArray(int size) {
            return new Psychotype[size];
        }
    };
}

File: app/src/main/java/com/laushkina/anastasia/psychosophy/domain/Psychotype.kt
package com.laushkina.anastasia.psychosophy.domain

import android.os.Parcel
import android.os.Parcelable
import com.laushkina.anastasia.psychosophy.domain.test.PsychoFunction

enum class Psychotype(val functions: Array<PsychoFunction>) : Parcelable {
    Augustine(arrayOf(PsychoFunction.Logic, PsychoFunction.Emotion, PsychoFunction.Physics, PsychoFunction.Will)),
    Pascal(arrayOf(PsychoFunction.Logic, PsychoFunction.Emotion, PsychoFunction.Will, PsychoFunction.Physics)),
    Berthier(arrayOf(PsychoFunction.Logic, PsychoFunction.Physics, PsychoFunction.Emotion, PsychoFunction.Will)),
    Plato(arrayOf(PsychoFunction.Logic, PsychoFunction.Physics, PsychoFunction.Will, PsychoFunction.Emotion)),
    Lao(arrayOf(PsychoFunction.Logic, PsychoFunction.Will, PsychoFunction.Physics, PsychoFunction.Emotion)),
    Einstein(arrayOf(PsychoFunction.Logic, PsychoFunction.Will, PsychoFunction.Emotion, PsychoFunction.Physics)),

    Anderson(arrayOf(PsychoFunction.Emotion, PsychoFunction.Logic, PsychoFunction.Will, PsychoFunction.Physics)),
    Rousseau(arrayOf(PsychoFunction.Emotion, PsychoFunction.Logic, PsychoFunction.Physics, PsychoFunction.Will)),
    Bukharin(arrayOf(PsychoFunction.Emotion, PsychoFunction.Physics, PsychoFunction.Logic, PsychoFunction.Will)),
    Pushkin(arrayOf(PsychoFunction.Emotion, PsychoFunction.Physics, PsychoFunction.Will, PsychoFunction.Logic)),
    Ghazali(arrayOf(PsychoFunction.Emotion, PsychoFunction.Will, PsychoFunction.Logic, PsychoFunction.Physics)),
    Parsnip(arrayOf(PsychoFunction.Emotion, PsychoFunction.Will, PsychoFunction.Physics, PsychoFunction.Logic)),

    Aristippus(arrayOf(PsychoFunction.Physics, PsychoFunction.Logic, PsychoFunction.Will, PsychoFunction.Emotion)),
    Epicurus(arrayOf(PsychoFunction.Physics, PsychoFunction.Logic, PsychoFunction.Emotion, PsychoFunction.Will)),
    Borgia(arrayOf(PsychoFunction.Physics, PsychoFunction.Emotion, PsychoFunction.Logic, PsychoFunction.Will)),
    Dumas(arrayOf(PsychoFunction.Physics, PsychoFunction.Emotion, PsychoFunction.Will, PsychoFunction.Logic)),
    Goethe(arrayOf(PsychoFunction.Physics, PsychoFunction.Will, PsychoFunction.Logic, PsychoFunction.Emotion)),
    Chekhov(arrayOf(PsychoFunction.Physics, PsychoFunction.Will, PsychoFunction.Emotion, PsychoFunction.Logic)),

    Akhmatova(arrayOf(PsychoFunction.Will, PsychoFunction.Emotion, PsychoFunction.Logic, PsychoFunction.Physics)),
    Tolstoy(arrayOf(PsychoFunction.Will, PsychoFunction.Emotion, PsychoFunction.Physics, PsychoFunction.Logic)),
    Lenin(arrayOf(PsychoFunction.Will, PsychoFunction.Logic, PsychoFunction.Physics, PsychoFunction.Emotion)),
    Socrates(arrayOf(PsychoFunction.Will, PsychoFunction.Logic, PsychoFunction.Emotion, PsychoFunction.Physics)),
    Napoleon(arrayOf(PsychoFunction.Will, PsychoFunction.Physics, PsychoFunction.Logic, PsychoFunction.Emotion)),
    Twardowski(arrayOf(PsychoFunction.Will, PsychoFunction.Physics, PsychoFunction.Emotion, PsychoFunction.Logic));

    companion object {
        private val DEFAULT = Augustine

        @JvmField
        val CREATOR = object : Parcelable.Creator<Psychotype> {
            override fun createFromParcel(parcel: Parcel): Psychotype {
                return resolve(parcel) ?: DEFAULT
            }

            override fun newArray(size: Int): Array<Psychotype?> {
                return arrayOfNulls(size)
            }
        }

        fun resolve(input: Parcel): Psychotype? {
            val parcelType = input.readString()
            return if (parcelType != null) valueOf(parcelType) else null
        }

        fun resolve(functions: Array<PsychoFunction>): Psychotype? {
            var foundType: Psychotype? = null
            for (psychotype in values()) {
                if (areFunctionsEqual(functions, psychotype.functions)) {
                    foundType = psychotype
                    break
                }
            }
            return foundType
        }

        fun areFunctionsEqual(f1: Array<PsychoFunction>, f2: Array<PsychoFunction>): Boolean {
            return f1.size == 4 && f2.size == 4
                    && f1[0] === f2[0]
                    && f1[1] === f2[1]
                    && f1[2] === f2[2]
                    && f1[3] === f2[3]
        }
    }

    override fun writeToParcel(parcel: Parcel, flags: Int) {
        parcel.writeString(name)
    }

    override fun describeContents(): Int {
        return 0
    }
}